### PR TITLE
chore(RHTAPWATCH-101): add contributing guide

### PR DESCRIPTION
**Pull request by srivickynesh**
Add a file to guide developers with setting up their dev environments and explains procedures and code review guidelines.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ok  	github.com/redhat-appstudio/segment-bridge.git/querygen	0.002s
+ok  	github.com/redhat-appstudio/segment-bridge.git/queryprint	(cached)
+ok  	github.com/redhat-appstudio/segment-bridge.git/scripts	0.002s
+ok  	github.com/redhat-appstudio/segment-bridge.git/segment	(cached)
+ok  	github.com/redhat-appstudio/segment-bridge.git/stats	(cached)
+ok  	github.com/redhat-appstudio/segment-bridge.git/webfixture	(cached)
+```
+4. If you want to run tests for a specific package, you can do so by providing the package path, like this:
+```
+go test github.com/redhat-appstudio/segment-bridge.git/querygen
```
Comment (amisstea) [inline CONTRIBUTING.md:32]: Or more simply:
```suggestion
go test ./querygen
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+### Running the Automated Tests
+[TBD]
+
+### Test Coverage
+[TBD]
+
+### Commit Messages
+We use [gitlint](https://jorisroovers.com/gitlint/) to standardize commit messages, following the [Conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format.
+
+If you include a Jira ticket identifier (e.g., STONEO11Y-123) in the commit message, PR name, or branch name, it will link to the Jira ticket.
```
Comment (amisstea) [inline CONTRIBUTING.md:47]:
```suggestion
If you include a Jira ticket identifier (e.g., RHTAPWATCH-123) in the commit message, PR name, or branch name, it will link to the Jira ticket.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    properly addressed.
+  * If a comment was addressed by the PR author but the reviewer did not resolve or
+    reply within 1 workday (reviewer's workday), then the comment can be resolved by
+    the PR author or by another reviewer.
+
+* All new and existing automated tests should pass.
+
+* A PR should be open for at least 1 workday at all time zones within the team. i.e.
+team members from all time zones should have an opportunity to review the PR within
+their workweek and their working hours.
```
Comment (amisstea) [inline CONTRIBUTING.md:77]:
```suggestion
their working hours.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+### Prerequisites
+* Basic tools: `curl`, `jq`, `oc`.
+* go: v1.20
+* Container engine: either `docker` or `podman`.
+
+### Unit Tests
+Go unit tests are included in various packages within the repository.
+Go unit tests are located within the tests directory, with filenames ending with _tests.go and with .go.
+
+### Running the Unit Tests Locally
```
Comment (amisstea) [inline CONTRIBUTING.md:16]:
```suggestion
#### Running the Unit Tests Locally
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+4. If you want to run tests for a specific package, you can do so by providing the package path, like this:
+```
+go test github.com/redhat-appstudio/segment-bridge.git/querygen
+```
+
+### Automated Tests
+[TBD]
+
+### Running the Automated Tests
+[TBD]
```
Comment (amisstea) [inline CONTRIBUTING.md:39]:
```suggestion
### Integration Tests
[TBD]

#### Running the Integration Tests
[TBD]
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+### Test Coverage
+[TBD]
```
Comment (amisstea) [inline CONTRIBUTING.md:42]: This should probably be moved under the unit test section.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+### Commit Messages
+We use [gitlint](https://jorisroovers.com/gitlint/) to standardize commit messages, following the [Conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format.
+
+If you include a Jira ticket identifier (e.g., STONEO11Y-123) in the commit message, PR name, or branch name, it will link to the Jira ticket.
+
+```
+feat(RHTAPWATCH-321): add new metrics
+
+- Added a new rule for tracking network transmit in bytes.
+- Included tests to verify the new metric.
```
Comment (amisstea) [inline CONTRIBUTING.md:53]: This example commit message is unrelated to the code in this repo. Let's pick a good one from our commit history instead.

**Review by hmariset — COMMENTED** [review]
other than Alex's comments everything else looks good.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+[TBD]
+
+### Running the Automated Tests
+[TBD]
+
+### Test Coverage
+[TBD]
+
+### Commit Messages
+We use [gitlint](https://jorisroovers.com/gitlint/) to standardize commit messages, following the [Conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format.
```
Comment (yftacherzog) [inline CONTRIBUTING.md:45]:
```suggestion
We use [gitlint](https://jorisroovers.com/gitlint/) to standardize commit messages, following the
[Conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+- Added a new rule for tracking network transmit in bytes.
+- Included tests to verify the new metric.
+
+Signed-off-by: Your Name <your-email@example.com>
+
+```
+
+### Pull Request Description
+When creating a Pull Request (PR), use the commit message as a starting point, and add a brief explanation. Include what changes you made and why. This helps reviewers understand your work without needing to investigate deeply. Clear information leads to a smoother review process.
```
Comment (yftacherzog) [inline CONTRIBUTING.md:60]:
```suggestion
When creating a Pull Request (PR), use the commit message as a starting
point, and add a brief explanation. Include what changes you made and why.
This helps reviewers understand your work without needing to investigate
deeply. Clear information leads to a smoother review process.
```

File: CONTRIBUTING.md (hunk 2 above)
Comment (Kousalya1998) [inline CONTRIBUTING.md:47]: Can you please provide an example of a correct commit message?

File: CONTRIBUTING.md (hunk 9 above)
Comment (Kousalya1998) [inline CONTRIBUTING.md:60]: As an alternative, we can create a PR description template similar to the one found in gitlab's app-interface.

- What?
- Why?
- Tickets
- Validation
- Dependency

Reference MR description: https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/77296

**Review by Kousalya1998 — CHANGES_REQUESTED** [review]

**Comment by ifireball** [comment]
Waiting with review until all pending comments are taken care of

File: CONTRIBUTING.md (hunk 9 above)
Comment (srivickynesh) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
+# Contributing
+
+This document provides guidelines for contributing to this repository.
+
+## Development Environment Setup
+
+### Prerequisites
+* Basic tools: `curl`, `jq`, `oc`.
+* go: v1.20
+* Container engine: either `docker` or `podman`.
```
Comment (ifireball) [inline CONTRIBUTING.md:10]: Only podman - we're not going to bother supporting both.

**Review by ifireball — CHANGES_REQUESTED** [review]
Missing section about how to build and run the container images for local testing - and how to run the scripts with them.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+## Development Environment Setup
+
+### Prerequisites
+* Basic tools: `curl`, `jq`, `oc`
+* go: v1.19
+* Container engine: `podman`
+
+### Setting up the Environment
+For initial setup, please refer to the
+[Running a Test Environment](./README.md#running-a-test-environment) section
```
Comment (ifireball) [inline CONTRIBUTING.md:14]: now that we have this document, that section belongs over here - and we should have a reference there

Comment (srivickynesh) [reply CONTRIBUTING.md:14]: Done

Comment (amisstea) [reply CONTRIBUTING.md:14]: @srivickynesh did you forget to push your changes? That section is still in the README.

**Review by Kousalya1998 — APPROVED** [review]
lgtm

File: CONTRIBUTING.md (hunk 11 above)
Comment (srivickynesh) [reply CONTRIBUTING.md]: Done. Pushed the correct commit

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+## Development Environment Setup
+
+### Prerequisites
+* Basic tools: `curl`, `jq`, `oc`
+* go: v1.19
+* Container engine: `podman`
+
+## Running a test environment
+
+### Kwok Container
```
Comment (ifireball) [inline CONTRIBUTING.md:14]: this section seems to be out of data - can you update it for building using our own custom image?

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+oc config set-context kwok --cluster=kwok
+```
+4. Set the Kwok context as the current context, if you've previously switched to another cluster, with the following command:
+```
+oc config use-context kwok
+```
+
+Now you can access the cluster using kubectl, e.g.: `kubectl get ns`.
+
+### Setting up a containerized Splunk instance
```
Comment (ifireball) [inline CONTRIBUTING.md:43]: Likewise, this should be updated to build and run our image rather then a generic one.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    to the [Splunk documentation][CS1].
+2. Once the container is running, you can log in to the Splunk instance using
+   the username `admin` and the password you set up.
+3. To access the Splunk [REST API][CS2],
+   you can make API calls from outside the container using the `curl` command.
+   For example, you can run the following command to search all data:
+     ```
+     curl -u admin:YourPassword -k https://localhost:8089/services/search/jobs -d search="search *"
+     ```
+4. If you want to use the Splunk UI, open a web browser on the host and navigate to
```
Comment (ifireball) [inline CONTRIBUTING.md:65]: There seems to be an unclosed quote somewhere, because syntax highlighting seems broken from this point on

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+ok  	github.com/redhat-appstudio/segment-bridge.git/segment	(cached)
+ok  	github.com/redhat-appstudio/segment-bridge.git/stats	(cached)
+ok  	github.com/redhat-appstudio/segment-bridge.git/webfixture	(cached)
+```
+4. If you want to run tests for a specific package, you can do so by providing
+the package path, like this:
+```
+go test ./querygen
+```
+
```
Comment (ifireball) [inline CONTRIBUTING.md:127]: if we don't have anything to write for these chapters, please do no include them until we do.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+#### Test Coverage
+[TBD]
+
+### Integration Tests
+[TBD]
+
+#### Running the Integration Tests
+[TBD]
+
+### Commit Messages
```
Comment (ifireball) [inline CONTRIBUTING.md:137]: please mention somewhere that one should run gofmt and all the unit tests before sending a PR.

File: CONTRIBUTING.md (hunk 13 above)
Comment (srivickynesh) [reply CONTRIBUTING.md:43]: @ifireball should this section use podman play kube or with podman build and run? what the preference here.

Comment (ifireball) [reply CONTRIBUTING.md:43]: whatever makes the life of the reader easier

File: CONTRIBUTING.md (hunk 15 above)
Comment (srivickynesh) [reply CONTRIBUTING.md:127]: I see there are tests for this that why I have included here as an example.

File: CONTRIBUTING.md (hunk 16 above)
Comment (srivickynesh) [reply CONTRIBUTING.md:155]: Done

File: CONTRIBUTING.md (hunk 13 above)
Comment (srivickynesh) [reply CONTRIBUTING.md:59]: Done

File: CONTRIBUTING.md (hunk 12 above)
Comment (srivickynesh) [reply CONTRIBUTING.md:14]: Done

**Review by ifireball — APPROVED** [review]
LGTM but someone who's not closely familiar with the repo should confirm this is clear enough and easy to follow

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,208 @@
+# Contributing
+
+This document provides guidelines for contributing to this repository.
+
+## Development Environment Setup
+
+### Prerequisites
+* Basic tools: `curl`, `jq`, `oc`
+* go: v1.19
+* Container engine: `podman`
```
[the remark this answers is not in the document]
Comment (ifireball) [reply CONTRIBUTING.md:10]: no we have code explicitly running `podman`, not going to bother making it portable.

[the remark this answers is not in the document]
Comment (ifireball) [reply CONTRIBUTING.md:10]: and we do use podman specific features - its not just a matter of the executable.

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,208 @@
+# Contributing
+
+This document provides guidelines for contributing to this repository.
+
+## Development Environment Setup
+
+### Prerequisites
+* Basic tools: `curl`, `jq`, `oc`
+* go: v1.19
```
[the remark this answers is not in the document]
Comment (srivickynesh) [reply CONTRIBUTING.md:9]: Go tests are written with v1.19 and some functions might not be compatible with higher versions while running the tests.

[the remark this answers is not in the document]
Comment (ifireball) [reply CONTRIBUTING.md:9]: its also what `go.mod` says.

[the remark this answers is not in the document]
Comment (ifireball) [reply CONTRIBUTING.md:9]: Since having go installed is a requirement for parsing go.mod (In other words, unlike with `pipenv` we don't have a tool to automatically get the right Go version on demand according to a file in the repo), we should specify it here so users will know what to install.

And yes, a basic requirements doc typically needs to get updates when you change the requirements...

**Comment by avi-biton** [comment]
LGTM

**Review by Omeramsc — APPROVED** [review]
LGTM
